man/whatsup.1
\."#################################################################
\."$Id: whatsup.1,v 1.4 2003-03-11 22:28:15 achu Exp $
\."by Albert Chu <chu11@llnl.gov>
\."#################################################################
.\"
.TH Whatsup 1 "Release 1.0" "LLNL" "Whatsup"
.SH NAME
whatsup \- list up and/or down nodes in a cluster
.SH SYNOPSIS
.B whatsup
[\fIOPTION\fR]... [\fINODES\fR]...
.br
.SH DESCRIPTION
.B whatsup
lists which nodes are currently up or down in a
cluster.  It determines which nodes are up or down based on
information gathered from genders and ganglia.

.B whatsup
will check on the status of every node in the cluster unless
specific nodes are listed at the command line.  The nodes can be 
listed in hostlist format, comma separated lists, or 
space separated lists.  See the section below on HOST RANGES
for instructions on how to list hosts in hostlist format.
What information is listed and how it
is listed is dependent on the options listed below.
.br
.SH OPTIONS
.TP
.B "-h, --help"
Print help and exit
.TP
.B "-V, --version"
Print version and exit
.TP
.B \-f \fISTRING\fR, \fB\-\-filename\fR=\fISTRING\fR
Location of genders file (default=/etc/genders)
.TP
.B \-o \fISTRING\fR, \fB\-\-hostname\fR=\fISTRING\fR
gmond server hostname (default=localhost)
.TP
.B \-i \fISTRING\fR, \fB\-\-ip\fR=\fISTRING\fR
gmond server IP address (default=127.0.0.1)
.TP
.B \-p \fIINT\fR, \fB\-\-port\fR=\fIINT\fR
gmond server port (default=8649)
.TP
.B \-b, \-\-updown
List both up and down nodes (default)
.TP
.B \-u, \-\-up
List only up nodes
.TP
.B \-d, \-\-down
List only down nodes
.TP
.B \-l, \-\-hostlist
List nodes in hostlist format (default)
.TP
.B \-c, \-\-comma
List nodes in comma separated list
.TP
.B \-n, \-\-newline
List nodes in newline separated list
.TP
.B \-s, \-\-space
List nodes in space separated list
.TP
.B \-a, \-\-altnames"
List nodes by alternate names (default=off)
.SH "HOST RANGES"
As noted in sections above,
.B whatsup
accepts ranges of hostnames in
the general form: prefix[n-m,l-k,...], where n < m and l < k, etc.,
as an alternative to explicit lists of hosts.  This form should not
be confused with regular expression character classes (also denoted
by ``[]''). For example, foo[19] does not represent foo1 or foo9, but
rather represents a degenerate range: foo19.

This range syntax is meant
only as a convenience on clusters with a prefixNN naming convention and
specification of ranges should not be considered necessary -- the list
foo1,foo9 could be specified as such, or by the range foo[1,9].

Some examples of range usage follow:

List all nodes among foo01,foo02,...,foo05 that are up and down 
    whatsup foo[01-05]

List all nodes among foo7,foo9,foo10 that are up and down
    whatsup foo[7,9-10]

As a reminder to the reader, some shells will interpret brackets
('[' and ']') for pattern matching.  Depending on your shell, it may 
be necessary to enclose hostlist formatted lists within quotes.  For 
example, the above two examples would be execuated as:

    whatsup "foo[01-05]"
    whatsup "foo[7,9-10]"

.SH "EXAMPLES"
.LP
List all nodes that are up and down.
     whatsup
.LP
List all nodes that are down
     whatsup -d
.LP
List all nodes among foo[0-20] that are down
     whatsup -d foo[0-20]
.LP
List all nodes that are up and down by their alternate names
     whatsup -a
.LP
List all nodes among foo[5-25] that are up by their alternate names
     whatsup -u -a foo[5-25]
.LP
List all nodes that are up in comma separated lists.	
     whatsup -u -c
.LP
List all nodes that are up among the nodes foo2, foo4, and foo8
     whatsup -u foo2, foo4, foo8
.SH "FILES"
/usr/bin/whatsup
.SH "SEE ALSO"
libgenders(3), libnodeupdown(3)
.SH "ORIGIN"
Developed by Albert Chu <chu11@llnl.gov> on LLNL's Linux 
clusters.  This software is open source and distributed under
the terms of the Gnu GPL.  


NEWS
This file describes changes in recent versions of whatsup. It
primarily documents those changes that are of interest to users
and admins.

Changes in whatsup-1.11-1
=========================
 -- Re-architect pingd daemon.
 -- Fix various minor typos and errors.

Changes in whatsup-1.10-1
=========================
 -- Support quick restarts on pingd server.
 -- Add documentation on configuration requirements.
 -- Fix manpage typos.
 -- If hostsfile is not found, assume default behavior instead of
    returning error.

Changes in whatsup-1.9-2
=========================
 -- Fix FC11/RHEL6 portability bugs.

Changes in whatsup-1.9-1
=========================
 -- Support different APIs to osm_mad_pool_init() for openib backend
    module.

Changes in whatsup-1.8-2
=========================
 -- Support building without perl extensions.

Changes in whatsup-1.8-1
===========================
- Various minor bug fixes.

Changes in whatsup-1.7-1
===========================
- Add libnodeupdown nodeupdown_last_up_time() function.
- Add last_up_time support to nodeupdown cerebro backend.
- Add whatsup -r option.

Changes in whatsup-1.6-1
===========================
- Add whatsup cerebro --monitor option.
- Add whatsup --log options.
- Add whatsup --module option.
- Add openib up/down backend module.

Changes in whatsup-1.5-1
===========================
- Added pingd daemon as possible whatsup backend tool.
- Support pingd in nodeupdown.

Changes in whatsup-1.4-1
===========================
- Whatsup -l option converted to -q option, for consistency with
genders 'nodeattr' command.
- Re-worked whatsup and nodeupdown library to use dynamically loadable
modules.
- Support Cerebro in nodeupdown.

Changes in whatsup-1.3-1
===========================
- Added nodeupdown_up_count and nodeupdown_down_count API functions.
- Added Nodeupdown and Libnodeupdown perl libraries.
- Added --count option to whatsup
- Added '-' stdin option.
- Output exit value based on whatsup results, allowsing whatsup to
  be used for scripts.
- Removed --ip option from whatsup
- Default whatsup output also includes output of node counts. 
- Whatsup and nodeupdown now compile and execute independently of 
  libganglia. 
- The nodeupdown.conf(5) file now allows specification of multiple
  hostnames instead of just one.
- Modified nodeupdown_load_data(3) API function to be far more
  generic and not based on libganglia.  This *will* break older code.
- To be consistent with changes to the API, nodeupdown.conf(5) no
  longer accepts the gmond_ip configuration option.
- Although the default is still genders, whatsup can now be compiled
  with different mechanisms for determining all of the nodes in a
  cluster.  See README file for details.

Changes in whatsup-1.2-1
===========================
 -- started NEWS file.

$Id: NEWS,v 1.20 2010-06-29 18:06:06 chu11 Exp $


whatsup.c
/*
 * $Id: whatsup.c,v 1.3 2003-02-21 01:24:25 achu Exp $
 * $Source: /g/g0/achu/temp/whatsup-cvsbackup/whatsup/whatsup.c,v $
 *    
 */

#include <errno.h>
#include <ganglia.h>
#include <genders.h>
#include <getopt.h>
#include <netdb.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/errno.h>
#include <sys/param.h>
#include <sys/socket.h>
#include <sys/types.h>

#ifdef WHATSUP_DEBUG
#include <mcheck.h>
#endif


#include "hostlist.h"
#include "nodeupdown.h"

/********************************
 * External Variables           *
 ********************************/

/* getopt */
extern char *optarg;
extern int optind, opterr, optopt;

/* gethostbyname */
extern int h_errno;

/********************************
 * Definitions                  *
 ********************************/

#define WHATSUP_OFF          0
#define WHATSUP_ON           1

#define WHATSUP_BUFFERLEN    65536

/* whatsup_output_type
 * - indicates if output should be nodes that are up or down
 */
enum whatsup_output_type {WHATSUP_UP, WHATSUP_DOWN};

/* whatsup_list_type
 * - indicates how output should be listed, in hostlist, comma separated, 
 *   neweline separated, or space separated listing.
 */
enum whatsup_list_type {WHATSUP_HOSTLIST, 
			WHATSUP_COMMA, 
			WHATSUP_NEWLINE, 
			WHATSUP_SPACE};

#ifdef WHATSUP_DEBUG

#define DEBUG_BUFFERLEN      65536
int debug_int;
int debug_buffer_len;
char *debug_buffer_ptr;

#endif

/* struct arginfo
 * - carries information about args passed in from the command line
 * genders_filename - filename of genders file
 * gmond_hostname - hostname of gmond server
 * gmond_ip - ip address of gmond server
 * gmond_port - port of gmond server
 * output_type - what output should be dumped to screen (up/down)
 * list_type - how nodes should be outputted (hostlist/comma/newline/space)
 * list_altnames - indicates if alternate names should be listed instead
 * nodes - stores the nodes (if any) were input in the command line
 */
struct arginfo {
  char *genders_filename; 
  char *gmond_hostname;   
  char *gmond_ip;         
  int gmond_port;         
  enum whatsup_output_type output_type;
  enum whatsup_list_type list_type;    
  int list_altnames;
  hostlist_t nodes;
};

/********************************
 * Function Declarations        *
 ********************************/

static void usage(void);
static void err_usage(char *);
static void output_error(char *, char *);
static void cleanup_struct_arginfo(struct arginfo *);
static int cmdline_parse(struct arginfo *, int, char **);
static int check_if_nodes_are_up_or_down(struct arginfo *, nodeupdown_t, hostlist_t);
static int get_up_or_down_nodes(struct arginfo *, nodeupdown_t, hostlist_t);
static int output_nodes(struct arginfo *, hostlist_t);

/* usage
 * - output usage and exit
 */
static void usage(void) {
  fprintf(stderr, "Usage: whatsup [OPTIONS] [nodes ...]\n");
  fprintf(stderr,"  -h        --help             Print help and exit\n");
  fprintf(stderr,"  -fSTRING  --filename=STRING  Location of genders file (default=%s)\n", DEFAULT_GENDERS_FILE);
  fprintf(stderr,"  -oSTRING  --hostname=STRING  gmond server hostname (default=localhost)\n");
  fprintf(stderr,"  -iSTRING  --ip=STRING        gmond server IP address (default=127.0.0.1)\n");
  fprintf(stderr,"  -pINT     --port=INT         gmond server port (default=%d)\n",GANGLIA_DEFAULT_XML_PORT);
  fprintf(stderr,"  -u        --up               List up nodes(default)\n");
  fprintf(stderr,"  -d        --down             List down nodes\n");
  fprintf(stderr,"  -l        --hostlist         List nodes in hostlist format (default)\n");
  fprintf(stderr,"  -c        --comma            List nodes in comma separated list\n");
  fprintf(stderr,"  -n        --newline          List nodes in newline separated list\n");
  fprintf(stderr,"  -s        --space            List nodes in space separated list\n");
  fprintf(stderr,"  -a        --altnames         List nodes by alternate names (default=off)\n");
  fprintf(stderr,"  nodes ...                    List of nodes to test (default=all noes)\n");
  fprintf(stderr,"\n");
  exit(1);
}

/* err_usage
 * - print error statement and output usage and exit
 */
static void err_usage(char *msg) {
  if (msg) {
    fprintf(stderr, "command line error: %s\n\n", msg);
  }
  usage();
}

/* output_error
 * - output error statement
 */
static void output_error(char *msg, char *errno_msg) {
  if (msg != NULL && errno_msg != NULL) {
    fprintf(stderr, "whatsup error: %s, %s\n", msg, errno_msg);
  }

  if (msg != NULL && errno_msg == NULL) {
    fprintf(stderr, "whatsup error: %s\n", msg);
  }

}

/* initialize_struct_arginfo
 * - initialize struct arginfo structure
 */
static int initialize_struct_arginfo(struct arginfo *arginfo) {
  arginfo->genders_filename = NULL;
  arginfo->gmond_hostname = NULL;
  arginfo->gmond_ip = NULL;
  arginfo->gmond_port = GANGLIA_DEFAULT_XML_PORT;
  arginfo->output_type = WHATSUP_UP;
  arginfo->list_type = WHATSUP_HOSTLIST;
  arginfo->list_altnames = WHATSUP_OFF;
  arginfo->nodes = NULL;
  return 0;
}

/* cleanup_struct_arginfo
 * - free memory allocated in a struct arginfo structure
 */
static void cleanup_struct_arginfo(struct arginfo *arginfo) {
  if (arginfo->genders_filename != NULL) {
    free(arginfo->genders_filename);
  }
  if (arginfo->gmond_hostname != NULL) {
    free(arginfo->gmond_hostname);
  }
  if (arginfo->gmond_ip != NULL) {
    free(arginfo->gmond_ip);
  }
  if (arginfo->nodes != NULL) {
    hostlist_destroy(arginfo->nodes);
  }
  free(arginfo);
}

/* cmdline_parse
 * - parse commandline arguments
 * - store info in a struct arginfo strcuture
 */
static int cmdline_parse(struct arginfo *arginfo, int argc, char **argv) {
  int hopt = 0, fopt = 0, oopt = 0, iopt = 0, popt = 0;
  int uopt = 0, dopt = 0, lopt = 0, copt = 0, nopt = 0, sopt = 0, aopt = 0;
  int c, index, i, ret;

  char *filename = DEFAULT_GENDERS_FILE;
  char *hostname = "localhost";
  char *ip = "127.0.0.1";
  int port = GANGLIA_DEFAULT_XML_PORT;

  char *options = "hf:o:i:p:udlcnsa";
  struct option long_options[] = {
    {"help",     0, NULL, 'h'},
    {"filename", 1, NULL, 'f'},
    {"hostname", 1, NULL, 'o'},
    {"ip",       1, NULL, 'i'},
    {"port",     1, NULL, 'p'},
    {"up",       0, NULL, 'u'},
    {"down",     0, NULL, 'd'},
    {"altnames", 0, NULL, 'a'},
    {"hostlist", 0, NULL, 'l'},
    {"comma",    0, NULL, 'c'},
    {"newline",  0, NULL, 'n'},
    {"space",    0, NULL, 's'},
    {0, 0, 0, 0}
  };

  /* turn off output messages printed by getopt_long */
  opterr = 0;

  while ((c = getopt_long(argc, argv, options, long_options, NULL)) != -1) {
    switch(c) {
    case 'h':
      hopt++;
      break;
    case 'f':
      fopt++;
      filename = optarg;
      break;
    case 'o':
      oopt++;
      hostname = optarg;
      break;
    case 'i':
      iopt++;
      ip = optarg;
      break;
    case 'p':
      popt++;
      port = atoi(optarg);
      break;
     case 'u':
      uopt++;
      break;
    case 'd':
      dopt++;
      break;
    case 'l':
      lopt++;
      break;
    case 'c':
      copt++;
      break;
    case 'n':
      nopt++;
      break;
    case 's':
      sopt++;
      break;
    case 'a':
      aopt++;
      break;
    case '?':
      err_usage("invalid command line option entered");
      break;
    default:
      output_error("getopt error", NULL);
      return -1;
      break;
    }
  }

  if (hopt > 0) {
    usage();
  }

  if (fopt > 1) {
    err_usage("you can only specify the --filename ('-f') option once");
  }

  if ((arginfo->genders_filename = strdup(filename)) == NULL) {
    output_error("out of memory", NULL);
    return -1;
  }

  if ((oopt + iopt) > 1) {
    err_usage("you can only specify one of the --gmond_hostname ('o') or --gmond_ip ('i') options");
  }
  else if (oopt == 1) {
    /* gmond hostname was provided, must also determine IP address */

    struct hostent *hptr;

    if ((arginfo->gmond_hostname = strdup(hostname)) == NULL) {
      output_error("out of memory", NULL);
      return -1;
    }
    
    if ((hptr = gethostbyname(hostname)) == NULL) {
      output_error("gethostbyname error", (char *)hstrerror(h_errno));
      return -1;
    }

    if ((arginfo->gmond_ip = (char *)malloc(INET_ADDRSTRLEN)) == NULL) {
      output_error("out of memory", NULL);
      return -1;
    }
    memset(arginfo->gmond_ip, '\0', INET_ADDRSTRLEN);

    if (inet_ntop(AF_INET, 
		  (void *)hptr->h_addr, 
		  arginfo->gmond_ip, 
		  INET_ADDRSTRLEN) == NULL) {
      output_error("inet_ntop error", strerror(errno));
      return -1;
    }
  }
  else if (iopt == 1) {
    /* gmond ip address was provided, must also determine hostname */

    struct hostent *hptr;
    struct in_addr temp_in_addr;
    int ret;

    if ((arginfo->gmond_ip = strdup(ip)) == NULL) {
      output_error("out of memory", NULL);
      return -1;
    }

    ret = inet_pton(AF_INET, arginfo->gmond_ip, &temp_in_addr);
    if (ret == 0) {
      output_error("inet_pton error", "invalid address");
      return -1;
    }
    else if (ret < 0) {
      output_error("inet_pton error", strerror(errno));
      return -1;
    }

    if ((hptr = gethostbyaddr(&temp_in_addr,
			      sizeof(struct in_addr), 
			      AF_INET)) == NULL) {
      output_error("gethostbyaddr error", (char *)hstrerror(h_errno));
      return -1;
    }

    if ((arginfo->gmond_hostname = strdup(hptr->h_name)) == NULL) {
      output_error("out of memory", NULL);
      return -1;
    }
  }
  else {
    /* use default hostname and ip address */

    if ((arginfo->gmond_hostname = strdup(hostname)) == NULL) {
      output_error("out of memory", NULL);
      return -1;
    }

    if ((arginfo->gmond_ip = strdup(ip)) == NULL) {
      output_error("out of memory", NULL);
      return -1;
    }
  }

  if (popt > 1) {
    err_usage("you can only specify --gmond_port ('-p') once");
  }
  else  if (popt == 1) {
    arginfo->gmond_port = port;
  }

  if ((uopt + dopt) > 1) {
    err_usage("you can only specify the --up ('-u') or --down ('-d') options once");
  }
  else if (uopt == 1) {
    arginfo->output_type = WHATSUP_UP;
  }
  else if (dopt == 1) {
    arginfo->output_type = WHATSUP_DOWN;
  }

  if ((lopt + copt + nopt + sopt) > 1) {
    err_usage("you can only specify one of the --hostlist ('-l'), --comma ('-c'), --newline ('-n'), or --space ('-s') options once");
  }
  else if (lopt == 1) {
    arginfo->list_type = WHATSUP_HOSTLIST;
  }
  else if (copt == 1) {
    arginfo->list_type = WHATSUP_COMMA;
  }
  else if (nopt == 1) {
    arginfo->list_type = WHATSUP_NEWLINE;
  }
  else if (sopt == 1) {
    arginfo->list_type = WHATSUP_SPACE;
  }

  if (aopt > 1) {
    err_usage("you can only specify --altnames ('-a') once");
  }
  else if (aopt == 1) {
    arginfo->list_altnames = WHATSUP_ON;
  }

  if ((arginfo->nodes = hostlist_create(NULL)) == NULL) {
      output_error("hostlist_create error", NULL);
      return -1;
  }

  /* store any nodes listed on the command line */
  index = optind;
  while (index < argc) {
    char *bracket1 = NULL;
    char *bracket2 = NULL;

    /* search for brackets.  Assume nodes are listed in hostlist format */

    bracket1 = strchr(argv[index], '[');
    bracket2 = strchr(argv[index], ']');
    
    if (bracket1 != NULL && bracket2 != NULL) {
      /* best situation, found both ends of a hostlist list of nodes */
      
      if ((ret = hostlist_push(arginfo->nodes, argv[index])) == 0) {
	output_error("hostlist_push error", 
		     "hosts may have been input incorrectly");
	return -1;
      }
    }
    else if (bracket1 != NULL && bracket2 == NULL) {
      /* hardest situation, found beginning bracket, but not end bracket */
      /* end of hostlist is in a later token */

      int index_begin = index;
      int strlen_count = strlen(argv[index]);
      char *buffer;
      index++;

      /* find ending token with end bracket */
      while (index < optind) {
	if ((bracket2 = strchr(argv[index], ']')) != NULL) {
	  break;
	}
	strlen_count += strlen(argv[index]);
	index++;
      }
      
      /* ending token not found */
      if (bracket2 == NULL) {
	output_error("'[' found, but not ']'", 
		     "hosts may have been input incorrectly");
	return -1;
      }

      /* create a single string out of all the tokens */
      buffer = (char *)malloc(strlen_count + 1);
      if (buffer == NULL) {
	output_error("out of memory", NULL);
	return -1;
      }
      memset(buffer, '\0', strlen_count + 1);
      
      strcpy(buffer, argv[index_begin]);
      for (i = (index_begin+1); i <= index; i++) {
	strcat(buffer, argv[i]);
      }

      if (hostlist_push(arginfo->nodes, buffer) == 0) {
	free(buffer);
	output_error("hostlist_push error", 
		     "hosts may have been input incorrectly");
	return -1;
      }

      free(buffer);
    }
    else if (bracket1 == NULL && bracket2 != NULL) {
      /* error situation, found end bracket before begin bracket */

      output_error("']' found before '['", 
		   "hosts may have been input incorrectly");
      return -1;
    }
    else {
      /* found no brackets, node input is comman or space separated */

      char *temp_char;

      if (strchr(argv[index],',') != NULL) {
	while ((temp_char = strchr(argv[index],',')) != NULL) {
	  *temp_char = ' ';
	}

	temp_char = strtok(argv[index]," ");
	while (temp_char != NULL) {
	  if (hostlist_push_host(arginfo->nodes, temp_char) == 0) {
	    output_error("hostlist_push_host error", 
			 "hosts may have been input incorrectly");
	    return -1;
	  }
	  temp_char = strtok(NULL," ");
	}
      }
      else {
	if (hostlist_push_host(arginfo->nodes, argv[index]) == 0) {
	  output_error("hostlist_push_host error", 
		       "hosts may have been input incorrectly");
	  return -1;
	}
      }
    }

    index++;
  }

  /* remove any duplicate nodes listed */
  hostlist_uniq(arginfo->nodes);

#ifdef WHATSUP_DEBUG
  fprintf(stderr,"arginfo\n");
  fprintf(stderr,"-------\n");
  fprintf(stderr,"genders_filename: %s\n", arginfo->genders_filename);
  fprintf(stderr,"gmond_hostname: %s\n", arginfo->gmond_hostname);
  fprintf(stderr,"gmond_ip: %s\n", arginfo->gmond_ip);
  fprintf(stderr,"gmond_port: %d\n", arginfo->gmond_port);
  fprintf(stderr,"output_type: %d\n", arginfo->output_type);
  fprintf(stderr,"list_type: %d\n", arginfo->list_type); 
  fprintf(stderr,"list_altnames: %d\n", arginfo->list_altnames);
  fprintf(stderr,"number of nodes: %d\n", hostlist_count(arginfo->nodes));

  debug_buffer_len = DEBUG_BUFFERLEN;
  if ((debug_buffer_ptr = (char *)malloc(debug_buffer_len)) == NULL) {
    output_error("out of memory", NULL);
    return -1;
  }
  memset(debug_buffer_ptr, '\0', debug_buffer_len);

  /* loop until the buffer is large enough */
  while (hostlist_ranged_string(arginfo->nodes, 
				debug_buffer_len, 
				debug_buffer_ptr) == -1) {
    free(debug_buffer_ptr);
    debug_buffer_len += DEBUG_BUFFERLEN;
    if ((debug_buffer_ptr = (char *)malloc(debug_buffer_len)) == NULL) {
      output_error("out of memory", NULL);
      return -1;
    }
    memset(debug_buffer_ptr, '\0', debug_buffer_len);
  }
  
  fprintf(stderr,"nodes inputted: %s\n", debug_buffer_ptr);
  free(debug_buffer_ptr);
  
  fprintf(stderr,"\n");
#endif

  return 0;
}

/* check_if_nodes_are_up_or_down
 * - determine if specific nodes passed in at the command line are up or down 
 */
int check_if_nodes_are_up_or_down(struct arginfo *arginfo, 
				  nodeupdown_t handle, 
				  hostlist_t nodes) {
  hostlist_iterator_t iter;
  char *str;
  int ret;

  if ((iter = hostlist_iterator_create(arginfo->nodes)) == NULL) {
    output_error("hostlist_iterator_create", NULL);
    return -1;
  }

  while ((str = hostlist_next(iter)) != NULL) {
    if (arginfo->output_type == WHATSUP_UP) {
      ret = nodeupdown_is_node_up(handle, str);
    }
    else {
      ret = nodeupdown_is_node_down(handle, str);
    }

    if (ret == 1) {
      if (hostlist_push_host(nodes, str) == 0) {
	hostlist_iterator_destroy(iter);
	output_error("hostlist_push_host", NULL);
	return -1;
      }
    }
    else if (ret == -1) {
      hostlist_iterator_destroy(iter);
      output_error("nodeupdown_is_node_up/down",
		   nodeupdown_strerror(nodeupdown_errnum(handle))); 
      return -1;
    }
			
  }
  hostlist_iterator_destroy(iter);

  return 0;
}

/* get_up_or_down_nodes
 * - get all up or down nodes
 */
int get_up_or_down_nodes(struct arginfo *arginfo, 
			 nodeupdown_t handle, 
			 hostlist_t nodes) {

  if (arginfo->output_type == WHATSUP_UP) {
    if (nodeupdown_get_up_nodes_hostlist(handle, nodes) == -1) {
      output_error("nodeupdown_get_up_nodes_hostlist",
		   nodeupdown_strerror(nodeupdown_errnum(handle)));
      return -1;
    }
  }
  else {
    if (nodeupdown_get_down_nodes_hostlist(handle, nodes) == -1) {
      output_error("nodeupdown_get_up_nodes_hostlist",
		   nodeupdown_strerror(nodeupdown_errnum(handle)));
      return -1;
    }
  }

  return 0;
}

/* output_nodes
 * - output the nodes indicated in nodes
 */
int output_nodes(struct arginfo *arginfo, hostlist_t nodes) {

  char *str = NULL;
  int i, str_len, num_nodes;
  char break_type;
  hostlist_iterator_t iter;

  /* sort nodes ahead of time */
  hostlist_sort(nodes);

  if (arginfo->list_type == WHATSUP_HOSTLIST) {
    /* output nodes in hostlist format */

    str_len = NODEUPDOWN_BUFFERLEN;
    if ((str = (char *)malloc(str_len)) == NULL) {
      output_error("out of memory", NULL);
      return -1;
    }
    memset(str, '\0', str_len);

    while (hostlist_ranged_string(nodes, str_len, str) == -1) {
      free(str);
      str_len += NODEUPDOWN_BUFFERLEN;
      if ((str = (char *)malloc(str_len)) == NULL) {
	output_error("out of memory", NULL);
	return -1;
      }
      memset(str, '\0', str_len);
    }

    fprintf(stdout, "%s\n", str);
    free(str);
    
  }
  else {
    /* output nodes separated by some break type */
    
    switch(arginfo->list_type) {
    case WHATSUP_COMMA:
      break_type = ',';
      break;
    case WHATSUP_NEWLINE:
      break_type = '\n';
      break;
    case WHATSUP_SPACE:
      break_type = ' ';
      break;
    default:
      output_error("output_nodes", "invalid list type");
      return -1;
      break;
    }

    if ((num_nodes = hostlist_count(nodes)) < 0) {
      output_error("hostlist_count", NULL);
      return -1;
    }

    if (num_nodes > 0) {
      if ((iter = hostlist_iterator_create(nodes)) == NULL) {
	output_error("hostlist_iterator_create", NULL);
	return -1;
      }
     
      for (i = 0; i < (num_nodes - 1); i++) {
	str = hostlist_next(iter);
	fprintf(stdout, "%s%c",str, break_type);
	free(str);
      }
      str = hostlist_next(iter);
      fprintf(stdout, "%s",str);
      free(str);
    }
    fprintf(stdout,"\n");

    hostlist_iterator_destroy(iter);
  }

  return 0;
}

int main(int argc, char **argv) {
  struct arginfo *arginfo = NULL;
  nodeupdown_t handle = NULL;
  hostlist_t nodes = NULL;
  hostlist_t alternate_nodes = NULL;

#ifdef WHATSUP_DEBUG
  mtrace();
#endif

  if ((arginfo = (struct arginfo *)malloc(sizeof(struct arginfo))) == NULL) {
    output_error("out of memory", NULL);
    goto cleanup;
  }

  if (initialize_struct_arginfo(arginfo) != 0) {
    output_error("initialize_struct_arginfo", NULL);
    goto cleanup;
  }

  if (cmdline_parse(arginfo, argc, argv) != 0) {
    goto cleanup;
  }
  
  if ((handle = nodeupdown_create()) == NULL) {
    output_error("nodeupdown_create", NULL);
    goto cleanup;
  }

  if (nodeupdown_load_data(handle, 
			   arginfo->genders_filename, 
			   arginfo->gmond_hostname, 
			   arginfo->gmond_ip, 
			   arginfo->gmond_port) == -1) {
    goto cleanup;
  }

  if ((nodes = hostlist_create(NULL)) == NULL) {
    goto cleanup;
  }

  /* get nodes based on whether we want all nodes or we need to check
   * just some of them
   */

  if (hostlist_count(arginfo->nodes) > 0) {
    if (check_if_nodes_are_up_or_down(arginfo, handle, nodes) != 0) {
      goto cleanup;
    }
  }
  else {
    if (get_up_or_down_nodes(arginfo, handle, nodes) != 0) {
      goto cleanup;
    }
  }

  /* get alternate names if we should output those instead */

  if (arginfo->list_altnames == WHATSUP_ON) {
    if ((alternate_nodes = hostlist_create(NULL)) == NULL) {
      output_error("hostlist_create", NULL);
      goto cleanup;
    }
    
    if (nodeupdown_get_hostlist_alternate_names(handle, 
						nodes, 
						alternate_nodes)) {
      output_error("nodeupdown_get_hostlist_alternate_names", 
		   nodeupdown_strerror(nodeupdown_errnum(handle)));
      goto cleanup;
    }

    if (output_nodes(arginfo, alternate_nodes) != 0) {
      goto cleanup;
    }
  }
  else {
    if (output_nodes(arginfo, nodes) != 0) {
      goto cleanup;
    }
  }

 cleanup:
  if (arginfo != NULL) {
    cleanup_struct_arginfo(arginfo);
  }
  
  if (handle != NULL) {
   (void)nodeupdown_destroy(handle);
  }

  if (nodes != NULL) {
    hostlist_destroy(nodes);
  }

  if (alternate_nodes != NULL) {
    hostlist_destroy(alternate_nodes);
  }

  exit(1);
}
